docs/freqai-feature-engineering.md

```diff
@@ -209,57 +209,77 @@ Another example, where the user wants to use live metrics from the trade databas
+FreqAI allows you to set a `weight_factor` to weight recent data more strongly than past data via an exponential function:
 
-## Data dimensionality reduction with Principal Component Analysis
+$$ W_i = \exp(\frac{-i}{\alpha*n}) $$
 
-You can reduce the dimensionality of your features by activating the `principal_component_analysis` in the config:
+where $W_i$ is the weight of data point $i$ in a total set of $n$ data points. Below is a figure showing the effect of different weight factors on the data points in a feature set.
 
-```json
-    "freqai": {
-        "feature_parameters" : {
-            "principal_component_analysis": true
-        }
-    }
-```
+![weight-factor](assets/freqai_weight-factor.jpg)
 
-This will perform PCA on the features and reduce their dimensionality so that the explained variance of the data set is >= 0.999. Reducing data dimensionality makes training the model faster and hence allows for more up-to-date models. 
+# Building the data pipeline
 
-## Inlier metric
+FreqAI uses the the [`DataSieve`](https://github.com/emergentmethods/datasieve) pipeline, which follows the SKlearn pipeline API, but adds, among other features, coherence between the X, y, and sample_weight vector point removals, and feature removal feature name following. 
 
-The `inlier_metric` is a metric aimed at quantifying how similar the features of a data point are to the most recent historical data points. 
+This means that users can use/customize any SKLearn modules and easily add them to their FreqAI data pipeline. By default, FreqAI builds the following pipeline:
 
-You define the lookback window by setting `inlier_metric_window` and FreqAI computes the distance between the present time point and each of the previous `inlier_metric_window` lookback points. A Weibull function is fit to each of the lookback distributions and its cumulative distribution function (CDF) is used to produce a quantile for each lookback point. The `inlier_metric` is then computed for each time point as the average of the corresponding lookback quantiles. The figure below explains the concept for an `inlier_metric_window` of 5.
+```py
+from datasieve.transforms import SKLearnWrapper, DissimilarityIndex
+from datasieve.pipeline import Pipeline
+dk.feature_pipeline = Pipeline([
+    ('scaler', SKLearnWrapper(MinMaxScaler(feature_range=(-1, 1))),
+    ('di', ds.DissimilarityIndex(di_threshold=1)),
+    ])
+```
 
-![inlier-metric](assets/freqai_inlier-metric.jpg)
+But users will find that they can add PCA and other steps just by changing their configuration settings, for example, if you add `"principal_component_analysis": true` to the `feature_parameters` dict in the `freqai` config, then FreqAI will add the PCA step for you resulting in the following pipeline:
+
+```py
+from datasieve.transforms import SKLearnWrapper, DissimilarityIndex, PCA
+from datasieve.pipeline import Pipeline
+dk.feature_pipeline = Pipeline([
```

Comment:
why would they (what's the usecase to set this directly)? (why do we need to have 2 interfaces for this?)

and in which method / part of which code would that code segment go? If we assume both interfaces as valid - then the docs will need some clarification for sure - as based on the code and descriptions here, i have no clue where this should be placed.

Reply:
Really? I am struggling to find where the misunderstanding is in the current docs:

```python
## Customizing the pipeline

Users are encouraged to customize the data pipeline to their needs by building their own data pipeline. This can be done by simply setting `dk.feature_pipeline` to their desired `Pipeline` object inside their `IFreqaiModel` `train()` function, or if they prefer not to touch the `train()` function, they can override `define_data_pipeline` in their `IFreqaiModel`:

```py
    from datasieve.transforms import SKLearnWrapper, DissimilarityIndex
    from datasieve.pipeline import Pipeline
    from sklearn.preprocessing import QuantileTransformer
    def define_data_pipeline(self) -> Pipeline:
        """
        User defines their custom eature pipeline here (if they wish)
        """
        feature_pipeline = Pipeline([
            ('qt', SKLearnWrapper(QuantileTransformer(output_distribution='normal'))),
            ('di', ds.DissimilarityIndex(di_threshold=1)
        ])

        return feature_pipeline
```

Here, you are defining the exact pipeline that will be used for your feature set during training and prediction. Here you can use *most* SKLearn transformation steps by wrapping them in the `SKLearnWrapper` class.

As there is the `feature_pipeline`, there also exists a definition for the `label_pipeline` which can be defined the same way as the `feature_pipeline`, by overriding `define_label_pipeline`.
```

Reply:
Yes - 3 is clearly too much.

the flags are fine - but i think we should focus on ONE method, and have only ONE of these methods defined.

I don't see a valid reason to have both - and it'll only be confusing for people.

Reply:
we should work based on the assumption that people do share their code / custom implementations of things we consider as "user facing".

Providing ONE clear way to do things will make it easier for people to understand what others did / tried to do, and improve based on that.

if we assume `train()` as user facing, then a common call to `define_data_pipeline()`  (with assignment to the dk object) is fine.

if that call to `define_data_pipeline()` is implemented by some users - but missing in favor of direct definition in other objects (created by other users, but shared), it'll make it VERY difficult for people to share, as it's 2 conflicting methods to do the same thing (If i define it directly, `define_data_pipeline ()` will not be called, which would be unexpected).

Reply:
Ok fixed, now the documentation highlights two methods for building a pipeline. From the config, or going in and customizing it yourself.